### PR TITLE
Full LS model

### PR DESCRIPTION
**Pull request by jeromekelleher**
This PR adds support for the full LS model with per-site mutation and recombination rates, and arbitrary mutations. This is not using the full algorithm being developed for tskit, but is importing the minimal changes required here to make the model work.

Still a WIP, will update when ready for review.

### CODE REVIEW

**Comment by hyanwong** [comment]
Fantastic. Look forward to reviewing it!

**Comment by jeromekelleher** [comment]
OK, I think this is ready for review @hyanwong. We're adding three new parameters to the various matching functions ``recombination_rate``, ``mutation_rate`` and ``precision``. These correspond directly to the values used in the underlying LS model. They're not population scaled values (no Ne). The ``precision`` controls the number of decimal places of accuracy that are used. This is an important parameter for performance, but the right value depends on the magnitudes of rho and mu.

I don't know how rho and mu should be interpreted --- this is to be figured out. The ``recombination_rate`` and ``mutation_rate`` parameters probably won't be what we settle on eventually as the right interface for tsinfer, but I don't know what that'll be yet. For now, these give us a way to have variable recombination rates along the genome, and to have some chance of having back mutations when we do sample matching. I'd suggest something like ``rho=1e-3``, mu=``1e-6`` as a starting point (so recombination is much more likely than mutation).

I think we should merge this much and start playing with these parameters to see what we get and build up some intuition on how they affect inference when we have error, etc. There's other improvements I want to make, but I want to improve the integration with tskit first so we're not doing all this messing around with translating coordinate systems, etc.

**Comment by hyanwong** [comment]
> I think we should merge this much and start playing with these parameters to see what we get and build up some intuition on how they affect inference when we have error, etc. 

Yes, this seems good to me, although I'm happy to start playing with this on a branch if we dan't want to merge it into master yet? But perhaps merging into master will force us to integrate with tskit more?

Speaking of which, it seems to me as if this PR doesn't use anything new from `tskit`, is that right? Or at least, I can't see any new python imports. Am I right in saying that the important changes are in the algorithm.py class, and its C equivalent, where you have switched to floating-point likelihoods (see my question in review about the main body of the algorithm).

**Comment by jeromekelleher** [comment]
> Yes, this seems good to me, although I'm happy to start playing with this on a branch if we dan't want to merge it into master yet? But perhaps merging into master will force us to integrate with tskit more?

I think we can merge. The default behaviour should be the same as before so we're not breaking anything. Performance will probably be significantly hit though, there's some tuning to be done to get the right default parameters.

> Speaking of which, it seems to me as if this PR doesn't use anything new from tskit, is that right? Or at least, I can't see any new python imports. Am I right in saying that the important changes are in the algorithm.py class, and its C equivalent, where you have switched to floating-point likelihoods (see my question in review about the main body of the algorithm).

That's right, we've back-ported logic from the tskit LS implementation to here.

I don't see any review questions: did you finish the review?

**Comment by hyanwong** [comment]
A further comment: an easy extra test would be to devise a simple example where the recombination breakpoint could parsimoniously be in one of two places, say between sites 1 and 2 or 2 and 3. If we set the site positions so that the distance between 1 & 2 is > that between 2 & 3, we should get recombination between 1 & 2. We can then reposition site 2 so that the smaller distance is between 2 & 3, and then the breakpoint should swap positions. Do you want me to code up a simple test case like this?

File: tsinfer/algorithm.py [hunk 1]
```diff
@@ -590,39 +594,21 @@ def dump_mutations(self):
...
-# currently not in the tree.
-
-
-MISMATCH = 0
-RECOMB = 1
-MATCH = 2
+# Special values used to indicate compressed paths and nodes that are
+# not present in the current tree.
 COMPRESSED = -1
 MISSING = -2
```
[the remark this answers is not in the document]
Comment (jeromekelleher) [reply tsinfer/algorithm.py]: Good call, changed to NONZERO_ROOT

File: tsinfer/algorithm.py [hunk 2]
```diff
@@ -808,8 +807,9 @@ def find_path(self, h, start, end, match):
         self.right_sib = np.zeros(n, dtype=int) - 1
         self.traceback = [{} for _ in range(m)]
         self.max_likelihood_node = np.zeros(m, dtype=int) - 1
+        self.allelic_state = np.zeros(n, dtype=int) - 1
 
-        self.likelihood = np.zeros(n, dtype=np.int8) - 2
+        self.likelihood = np.zeros(n) - 2
```
[the remark this answers is not in the document]
Comment (jeromekelleher) [reply tsinfer/algorithm.py]: True. I've changed the ``-2`` one, but it's too much work to go through all of them for the ``-1`` cases. This code will all be deleted soon, so not much point in polishing I think.

File: tsinfer/algorithm.py [hunk 3]
```diff
@@ -658,70 +645,82 @@ def check_likelihoods(self):
...
+                p_t = p_no_recomb
+            else:
+                p_t = p_recomb
+                recombination_required = True
+            self.traceback[site][u] = recombination_required
+            p_e = mu
+            if haplotype_state == self.allelic_state[v]:
+                # p_e = 1 - (len(alleles) - 1) * mu
+                p_e = 1 - mu
+            self.likelihood[u] = round(p_t * p_e, self.precision)
```
[the remark this answers is not in the document]
Comment (jeromekelleher) [reply tsinfer/algorithm.py]: Yes.

File: tests/test_inference.py [hunk 4]
```diff
@@ -304,20 +308,32 @@ def verify_data_round_trip(
...
+        if sequence_length is None:
+            sequence_length = positions[-1] + 1
+        with tsinfer.SampleData(sequence_length=sequence_length) as sample_data:
+            for j in range(genotypes.shape[0]):
+                t = None if times is None else times[j]
+                site_alleles = None if alleles is None else alleles[j]
+                sample_data.add_site(positions[j], genotypes[j], site_alleles, time=t)
+        ancestors = tsinfer.generate_ancestors(sample_data)
+        ancestors_ts = tsinfer.match_ancestors(sample_data, ancestors)
+        rho = [1e-9, 1e-3, 0.1]
```
[the remark this answers is not in the document]
Comment (jeromekelleher) [reply tests/test_inference.py]: Yes, good call. I've added checks for mu=0 and rho=0.

Capturing where inference fails needs some follow up work I think. I'll open an issue.

**Comment by jeromekelleher** [comment]
> A further comment: an easy extra test would be to devise a simple example where the recombination breakpoint could parsimoniously be in one of two places, say between sites 1 and 2 or 2 and 3. If we set the site positions so that the distance between 1 & 2 is > that between 2 & 3, we should get recombination between 1 & 2. We can then reposition site 2 so that the smaller distance is between 2 & 3, and then the breakpoint should swap positions. Do you want me to code up a simple test case like this?

Please do!

File: tsinfer/algorithm.py (hunk 2 above)
[the remark this answers is not in the document]
Comment (hyanwong) [reply tsinfer/algorithm.py]: > Please do!

Will do!

**Comment by hyanwong** [comment]
> Please do!

Will do

**Pull request merged**